_delb/nodes.py

```diff
@@ -3155,229 +3190,676 @@ def __collect_prefixes(self, root: TagNode):  # noqa: C901  REMOVE eventually
+            # and one would know that the line starts with text content
+            #
+            # so for now that's why text might start after a closing tag of an
+            # element that spans multiple lines on their last one
+            # (i.e. this prefers: "the <hi>A</hi> letter…" over the "A" on a
+            # separate line which the next branch produces)
+
+            if self._whitespace_is_legit_before_node(nodes[0]):
+                lines.append("")
+            lines.extend(self._wrap_text(content.lstrip(), self._width))
+
+        else:
+            if content.startswith(" "):
+                filling = " " + next(
+                    self._wrap_text(content[1:], self._available_space - 1)
+                )
+            else:
+                filling = next(self._wrap_text(content, self._available_space))
+
+            if not (
+                len(filling) > self._available_space
+                and self._whitespace_is_legit_before_node(nodes[0])
+            ):
+                self.writer(filling)
+                content = content[len(filling) + 1 :]
+                if not content:
+                    return
+
+            lines.append("")
+            lines.extend(self._wrap_text(content, self._width))
+
+        if (
+            lines[-1].endswith(" ")
+            and self._whitespace_is_legit_after_node(nodes[-1])
+            and (following_sibling := nodes[-1].fetch_following_sibling()) is not None
+            and not self._required_space(
+                following_sibling, self._width - len(lines[-1])
+            )
+        ):
+            lines.append("")
+
+        self._consolidate_text_lines(lines)
+
+        prefix = self._level * self.indentation
+        for line in lines[:-1]:
+            if line == "":
+                self.writer("\n")
+            else:
+                self.writer(f"{prefix}{line}\n")
+        if lines[-1] != "":
+            self.writer(f"{prefix}{lines[-1]}")
+
+    def _consolidate_text_lines(self, lines: list[str]):
+        last_node = self._unwritten_text_nodes[-1]
+        assert last_node.parent is not None
+        if (
+            lines[0] == ""
+            and last_node is last_node.parent.last_child
+            and self._whitespace_is_legit_after_node(last_node)
+        ):
+            lines.append("")
+        if self._line_offset == 0 and lines[0] == "":
+            lines.pop(0)
+        if len(lines) >= 2 and lines[-1] == "":
+            lines[-2] = lines[-2].rstrip()
+
+    @staticmethod
+    def _wrap_text(text: str, width: int) -> Iterator[str]:
+        while len(text) > width:
+
+            if (index := text.rfind(" ", 0, width + 1)) > -1 or (
+                index := text.find(" ", width)
+            ) > 0:
+                yield text[:index]
+                text = text[index + 1 :]
+            else:
+                yield text
+                return
+
+        if text:
+            yield text
+
+
+class PrettyFormatOptions(NamedTuple):
```

Comment:
i'm already beginning to abbreviate this as `pfo`. `FormatOptions` would suffice, wouldn't it? the relating arguments similarly.